src/puppetlabs/jdbc_util/core.clj

```diff
@@ -190,14 +190,21 @@
       :pg_get_serial_sequence))
 
 (defn reconcile-sequence-for-column!
-  "Finds the sequence associated with the given column and sets it to the
-  current maximum value in that column, so that the next value in the sequence
-  will be the current maximum + 1 (or 0, if the table is empty). If the column
-  has no associated sequence, throws an Exception."
+  "Finds the sequence associated with the given column and sets it
+  to the greatest of:
+    * the maximum value currently in the column + 1
+    * the current value of the sequence
+    * 1 (if the column is empty)
+  If the column has no associated sequence, throws an Exception."
   [db table column]
   (if-let [sequence-name (get-sequence-name-for-column db table column)]
-    (jdbc/with-db-transaction [txn-db db]
-      (jdbc/execute! txn-db [(format "LOCK TABLE \"%s\" IN EXCLUSIVE MODE" table)])
-      (jdbc/query txn-db [(format "SELECT setval('%s', COALESCE(max(\"%s\")+1, 1), false) FROM \"%s\""
-                                  sequence-name column table)]))
+    (let [select-last-value (str "(SELECT last_value FROM " sequence-name ")")
+          select-max-in-column (str "(SELECT MAX(" column ") FROM " table ")")]
+      (jdbc/with-db-transaction [txn-db db]
+        (jdbc/execute! txn-db [(format "LOCK TABLE \"%s\" IN EXCLUSIVE MODE" table)])
+        (jdbc/query txn-db [(str "SELECT setval('" sequence-name "',"
+                                 "              GREATEST(" select-last-value ","
+                                 "                       " select-max-in-column " + 1,"
+                                 "                       1),"
+                                 "              false)")])))
```

Comment:
In the case where the most recent rows have been deleted, this will cause the latest value to be repeated on the next insert. If rows 1, 2, 3, 4 are inserted and then 3 and 4 are deleted, `last_value` will be 4 and `is_called` will be true. After this code executes, `last_value` will still be 4, but `is_called` will be false. That will cause the next insert to use the value 4.

This needs to either use `last_value + 1` for the first case, or it needs to use `MAX()` for the second case and set `is_called` to true. But setting `is_called` to true is wrong if we're setting the value to its default of 1 because it has never been called (we're not allowed to set it to 0).

So why not just *not* call `setval()` if the current value is already correct?

```sql
SELECT CASE (SELECT MAX(id) FROM jobs) > SELECT last_value FROM jobs_id_seq THEN setval('job_id_seq', SELECT MAX(id) FROM jobs) ELSE END;
```